### PR TITLE
HDDS-9254. Legacy replication manager uses mismatched replicas as replication sources

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Consider a case where SCM has a CLOSED container and the replica states are CLOSED, CLOSED, QUASI. In the pull replication model, RM will send all 3 of these replicas to the datanode to use as replication sources. The DN will do a random shuffle and pick one to replicate. If it chooses the QUASI-CLOSED replica, the next iteration of RM will see replicas CLOSED, CLOSED, QUASI, QUASI. RM will issue the same command since the CLOSED replicas are still under replicated, but now the odds of the DN's random shuffle choosing a quasi closed replica are increased. This process can repeat until the cluster is filled with a quasi-closed replica on each datanode. This can bring the cluster into the stuck state described in [HDDS-8536](https://issues.apache.org/jira/browse/HDDS-8536).

## What is the link to the Apache JIRA

HDDS-9254

## How was this patch tested?

Unit tests added.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java [hunk 1]
```diff
@@ -1138,8 +1138,12 @@ private void handleUnderReplicatedHealthy(final ContainerInfo container,
           container.containerID());
     }
 
+    State matchingReplicaState = State.CLOSED;
+    if (container.getState() == LifeCycleState.QUASI_CLOSED) {
+      matchingReplicaState = State.QUASI_CLOSED;
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java:1143]: If we are in the container=QUASI_CLOSED branch, then do we need to ensure we only replicate the largest sequenceID quasi_closed replica? There isn't any logic in `getReplicationSources` to do that right now.

Comment (errose28) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java:1143]: This is not required. I added a comment to this part of the code to explain why, and added a unit test to verify this behavior. Let me know if the comment makes sense.

**Review by sodonnel — APPROVED** [review]
LGTM

**Comment by errose28** [comment]
@sodonnel I just resolved the merge conflict.

**Review by sodonnel — APPROVED** [review]

**Pull request merged**